core/state/trie_prefetcher.go

```diff
@@ -451,20 +452,21 @@ func (sf *subfetcher) loop() {
 					}
 				}
 			}
+			tasks = tasks[:0] // avoid reallocation
 
 		case <-sf.stop:
 			// Termination is requested, abort if no more tasks are pending. If
 			// there are some, exhaust them first.
-			sf.lock.Lock()
-			done := sf.tasks == nil
-			sf.lock.Unlock()
-
-			if done {
+			if len(tasks) > 0 {
+				// See earlier invariant that guarantees a receive on `work` to clear `tasks`.
+				continue
+			}
+			select {
+			case ts := <-sf.tasks:
+				addTasks(ts)
+			default:
```

Comment:
uu, a select in a closure in a select in a select

Reply:
Yeah, it was smelly!